geo_tag/geo_tag.py

```diff
@@ -0,0 +1,200 @@
+import json
+import os
+import pickle
+
+from geopy.geocoders import Nominatim
+
+class TwitterJSONTagger():
+    def __init__(self):
+        # data is the tweet json which we add geo_tag.
+        self.data = dict()
+        # cache is the hashmap { (city, state), other geo info}
+        self.cache = dict()
+        # coord is the (longitude, latitude)
+        self.coord = tuple()
+        # self.abbrev_us_state is the relation between the state's full name and abbrev.
+        with open('us_state_abbrev.json', 'r') as inf:
+            self.abbrev_us_state = eval(inf.read())
+
+    def init_geo_cache(self, city_json):
+        # check pickle module
+        filename = "cache_geo_hashmap"
+        if os.path.exists(filename):
+            cache_file = open(filename, 'rb')
+            self.cache = pickle.load(cache_file)
+            return
+        else:
+            feature = city_json.get('features')
+            for f in feature:
+                item = f.get('properties')
+                # store the city, state as key; (cityID, countyID, stateID, countyName) as value.
+                if item["name"] is not None and item["stateName"] is not None:
+                    _city_state = item["name"] + ", " + item["stateName"]
+                    # key, value
+                    self.cache[_city_state] = (item["cityID"], item["countyID"], item["stateID"], item["countyName"])
+
+                else:
+                    print("[ERROR] city.json data.")
+                    return
+
+            # store it using pickle
+            cache_file = open(filename, 'wb')
+            pickle.dump(self.cache, cache_file)
+
+    def get_coordinate(self):
+        coord = self.data.get('coordinates')
+        place = self.data.get('place')
+
+        if coord is not None:
+            self.coord = coord.get('coordinates')
+        else:
+            # 2. get the central point of the bounding_box
+            if place is not None:
+                bounding_box = place.get('bounding_box').get('coordinates')[0]
+                ''' Twitter has some wield format historically, though it still rectangle, 
+                but it is not always in (sw, se, ne, nw) order
+                '''
+                swLog = min(bounding_box[0][0], bounding_box[1][0], bounding_box[2][0], bounding_box[3][0])
+                swLat = min(bounding_box[0][1], bounding_box[1][1], bounding_box[2][1], bounding_box[3][1])
+                neLog = max(bounding_box[0][0], bounding_box[1][0], bounding_box[2][0], bounding_box[3][0])
+                neLat = max(bounding_box[0][1], bounding_box[1][1], bounding_box[2][1], bounding_box[3][1])
```

Comment:
have this chunk of code being tested? I am not checking details.

Reply:
Yep, tested. This logic is the same as the previous version in cloudberry which we talked before, and here is the [link](https://github.com/ISG-ICS/cloudberry/blob/4538dca3c2fae5c7334e9db02b918e5ae796cfcf/examples/twittermap/noah/src/main/scala/edu/uci/ics/cloudberry/noah/adm/ADM.java)

Reply:
ok. I rely on you for the details. Thanks.